net-im/toxic/toxic-0.8.0-r2.ebuild

```diff
@@ -0,0 +1,93 @@
+# Copyright 1999-2018 Gentoo Foundation
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=6
+PYTHON_COMPAT=( python3_4 python3_5 python3_6 )
+
+inherit python-single-r1 xdg
+
+DESCRIPTION="A curses-based client for Tox."
+HOMEPAGE="https://github.com/JFreegman/toxic"
+SRC_URI="https://github.com/JFreegman/toxic/archive/v${PV}.tar.gz -> ${P}.tar.gz"
+
+LICENSE="GPL-3"
+SLOT="0"
+KEYWORDS="~amd64 ~x86"
+IUSE="+X +audio notifications +python qrpng +video"
+REQUIRED_USE="python? ( ${PYTHON_REQUIRED_USE} )"
+
+# Not a typo; net-libs/tox only has a 'both or neither' option
+RDEPEND="
+	|| (
+	audio? ( net-libs/tox:0/0.1[av] )
+	video? ( net-libs/tox:0/0.1[av] )
+	net-libs/tox:0/0.1
+	)
+	dev-libs/libconfig
+	media-gfx/qrencode
+	media-libs/libpng:0=
```

Comment:
I think this should be USE-conditional, shouldn't it?

Reply:
This is why I omitted listing this dep originally. It's a hard dep of qrencode. Therefore, it may seem intuitive that you shouldn't need libpng to build this package, but because qrencode requirs it anyway, disabling the PNG support just affects the support within toxic itself only, and does not alleviate the need to have libpng to compile.

That reasoning is the entire basis of this revision, really. I had the same surprise myself when working with this package, which is why I went as far as reading the source code to figure out what's going on. The OP post contains references of my findings supporting the conclusion.

Reply:
You're going too far in context. If this package has optional PNG support, and you expose it via USE flag, you should also make the dependency optional. Otherwise, the package built with `USE=-qrpng` has a dependency on libpng even though it doesn't use it.

Of course, whether disabling PNG support makes any sense is another question. If it doesn't, you can remove the flag entirely and keep it always on.

Just don't do it halfway, please.

Reply:
Why does a USE flag have to be obligated to represent any number of dependencies, whether that is zero, one, or more? I thought they only represent build-time options, usually the sort you see as optional settings in a `configure` script. There really is an option in this project's build process to enable or disable the support for PNG in the QR part of its code. This USE flag is honestly representing what it claims to do. I would have preferred for this flag to be what the original flag suggests it to be -- disabling the QR support entirely. This project is not programmed to allow this and thus it is also the reason why I renamed the flag to eliminate potential deception about its purpose. You can't compile it without qrencode. I have tried doing exactly that to be certain of it, as well as reading the source code to verify this is why trying that failed and to properly grasp what exactly is being disabled when that option is shut off in the build process. As expected, that is indeed the state of it.

Also, no package is required to actually implement everything its dependencies offer, and I don't see why we should be stripping a USE flag over a misnomer (then again, maybe it isn't, due to the renaming of it.) We should be allowed to offer by proxy via USE build-time options so there is no penalty for installing via Portage rather than compiling manually. If you shut off the proposed `qrpng` flag, you disable Toxic's support for the PNG generation `qrencode` offers.

The presence of the hard dependency is self-explanatory: you need it to build this thing, or it won't work. So I see no need to further cater to users emerging this package via `elog` or anything of that nature, since the dep tree speaks for itself. But I can understand leaving a comment in the ebuild itself so other maintainers reading the ebuild understand the situation we're discussing here.

I'd like to be sure that the relationships of the dependencies here and the structure of Toxic itself are clarified before a conclusion is drawn.

As for whether having PNG support an option makes sense, I don't have a say in that; that's upstream's design of their program. I made the changes to mirror what their build process offers. "The system must provide library/feature X anyway, may as well forcefully enable X" sounds like a slippery slope to me. We have other packages with flags that probably fall into a very similar predicament, and I myself have likely disabled things on purpose nonetheless. If I were one of the devs working on Toxic, I'd probably change the program structure so that building without QR is possible, but right now, for version 0.8.0 at least, it isn't written that way.